server/src/main/java/org/elasticsearch/common/util/concurrent/AtomicArray.java

```diff
@@ -98,20 +100,20 @@ public List<E> asList() {
                         list.add(e);
                     }
                 }
-                nonNullList = list;
+                res = list;
             }
+            nonNullList = res;
         }
-        return nonNullList;
```

Comment:
This is a potential bug, if you update this thing concurrently `nonNullList` can be nulled out between the check above and returning.